sdk/eventhub/azure-eventhubs-aio-partitionmanager-blobstorage/azure/eventhub/aio/partitionmanager/blobstorage/blobstoragepm.py

```diff
@@ -0,0 +1,125 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+from typing import Iterable, Dict, Any
+import logging
+from collections import defaultdict
+import asyncio
+from azure.eventhub.eventprocessor import PartitionManager, OwnershipLostError
+from azure.core.exceptions import ResourceModifiedError, ResourceExistsError, AzureError
+from azure.storage.blob.aio import ContainerClient
+
+logger = logging.getLogger(__name__)
+UPLOAD_DATA = ""
+
+
+class BlobPartitionManager(PartitionManager):
+    """An PartitionManager that uses Azure Blob Storage to store the partition ownership and checkpoint data.
+
+    This class implements methods list_ownership, claim_ownership, and update_checkpoint that are defined in class
+    azure.eventhub.eventprocessor.PartitionManager of package azure-eventhub.
+
+    """
+    def __init__(self, container_client: ContainerClient):
+        """
+
+        :param container_client: The Azure Blob Storage Container client.
+        """
+        self._container_client = container_client
+        self._cached_ownership_dict = defaultdict(dict)  # type: Dict[str, Dict[str, Any]]
+        # lock each partition for list_ownership, claim_ownership and update_checkpoint etag doesn't get out of sync
+        # when the three methods are running concurrently
+        self._cached_ownership_locks = defaultdict(asyncio.Lock)
+
+    async def list_ownership(self, eventhub_name: str, consumer_group_name: str) -> Iterable[Dict[str, Any]]:
+        try:
+            blobs = self._container_client.list_blobs(include=['metadata'])
+        except AzureError as azure_err:  # list_blobs has exhausted retry
+            logger.exception("An exception occurred during list_ownership for eventhub %r consumer group %r."
+                             " An empty ownership list is returned",
+                             eventhub_name, consumer_group_name, exc_info=azure_err)
+            return []
+        async for b in blobs:  # TODO: running them concurrently
+            async with self._cached_ownership_locks[b.name]:
+                metadata = b.metadata
+                ownership = {
+                    "eventhub_name": eventhub_name,
+                    "consumer_group_name": consumer_group_name,
+                    "partition_id": b.name,
+                    "owner_id": metadata["owner_id"],
+                    "etag": b.etag,
+                    "last_modified_time": b.last_modified.timestamp() if b.last_modified else None
+                }
+                ownership.update(metadata)
```

Comment:
Do we really want metadata to override any properties previously set in case of name conflicts?

Reply:
Yes, metadata(owner_id, offset, sequence_number) should override.

Reply:
Just to be sure, if the blob has metadata `"partition_id": "..."`, you want "..." as the value for `"partition_id"` that rather than the name of the blob?

Reply:
Our code put and update what's in metadata. If that happens, there is already a bug causing corrupt data.
"partition_id" is not a part of metadata. It's used as the blob name.